src/pdf_beamtime/src/pdf_beamtime_server.cpp

```diff
@@ -247,6 +294,80 @@ void PdfBeamtimeServer::new_obstacle_service_cb(
   planning_scene_interface_.applyCollisionObjects(create_env());
 }
 
+int PdfBeamtimeServer::get_state_completions()
+{
+  return progress_ / total_states_;
+}
+
+bool PdfBeamtimeServer::run_fsm(
+  std::shared_ptr<const pdf_beamtime_interfaces::action::PickPlaceControlMsg_Goal> goal)
+{
+  RCLCPP_INFO(node_->get_logger(), "Executing state %d", static_cast<int>(current_state_));
+  bool state_transition = false;
+  switch (current_state_) {
+    case State::HOME:
+      state_transition = set_joint_goal(goal->pickup_approach);
+      break;
+
+    case State::PICKUP_APPROACH:
+      state_transition = set_joint_goal(goal->pickup);
+      break;
+
+    case State::PICKUP:
+      // gripper_close();
+      state_transition = true;
+      break;
+
+    case State::GRASP_SUCCESS:
+      state_transition = set_joint_goal(goal->pickup_approach);
+      break;
+
+    case State::PICKUP_RETREAT:
+      state_transition = set_joint_goal(goal->place_approach);
+      break;
+
+    case State::PLACE_APPROACH:
+      state_transition = set_joint_goal(goal->place);
+      break;
+
+    case State::PLACE:
+      // gripper_open();
```

Comment:
Add a TODO for checking the parameter server for "gripper_avail" or something to that effect?

Reply:
TODO added
https://github.com/maffettone]/erobs/blob/98d719dd09aa992e1f53f067c1cf9668e64f753c/src/pdf_beamtime/src/pdf_beamtime_server.cpp#L356